pythran/pythonic/numpy/fft/irfft.hpp

```diff
@@ -43,24 +43,39 @@ namespace numpy
 
       // Create the twiddle factors. These must be kept from call to call as
       // it's very wasteful to recompute them.
+      // The call to npy_irfftf makes changes to the twiddle factor buffer (then
+      // restores it) so it's
+      // not thread safe. To avoid this:
+      // I keep one copy of the twiddle factors that no thread is allowed to
+      // modify. Then each thread
+      // has its own copy of it that it can modify at will.
+      // This is from fftpack_litemodule.c
+      static std::map<long, std::vector<double>> all_twiddles_irfft_global;
+      static thread_local std::map<long, std::vector<double>> all_twiddles_irfft_local;
       mtx_irfft.lock();
-      static std::map<long, types::ndarray<double, types::pshape<long>>>
-          all_twiddles_irfft;
-      if (all_twiddles_irfft.find(NFFT) == all_twiddles_irfft.end()) {
+      if (all_twiddles_irfft_global.find(NFFT) ==
+          all_twiddles_irfft_global.end()) {
+        // This happens only once across all threads for a give NFFT
         // Insert a new twiddle array into our map and initialize it
-        all_twiddles_irfft.insert(std::make_pair(
-            NFFT, types::ndarray<double, types::pshape<long>>(
-                      {(long)(2 * NFFT + 15)}, __builtin__::None)));
-        npy_rffti(NFFT, (double *)all_twiddles_irfft[NFFT].buffer);
+        all_twiddles_irfft_global.emplace(NFFT, std::vector<double>(2 * NFFT + 15));
+        // std::cout << "INIT GLOBAL " << NFFT << " \n";
+        // This is the (costly) initialization of the array.
+        npy_rffti(NFFT, all_twiddles_irfft_global[NFFT].data());
       }
+
+      if (all_twiddles_irfft_local.find(NFFT) == all_twiddles_irfft_local.end()) {
+        // This happens once per thread.
+        all_twiddles_irfft_local.emplace(NFFT, std::vector<double>(all_twiddles_irfft_global[NFFT]));
```

Comment:
you don't need to call  `std::vector<double>(all_twiddles_irfft_global[NFFT])` here, `all_twiddles_irfft_global[NFFT]` should be enough. Note that if the global value doesn't change, you could store a pointer to it, avoiding an extra copy (like the global has ownership, and the local just stores aview (you could use a unique_pointer to reflect that, but dont worry with that). If the local needs a copy, then just removin the extra ``std::vector`` is ok.

Reply:
The local really needs a copy because otherwise it would modify the global and screw up everybody else! So when I call emplace(NFFT,existing_array) what's put in the map is a *copy* of existing_array? I'm not very clear on which situtations emplace avoids a copy. But I'll do what you recommend.